### PR TITLE
Fix local_work_size for col2im/im2col and layout transposing kernels

### PR DESCRIPTION
**Pull request by atamazov**
### Full blown fix for issue #2047.

OCL spec says (see clEnqueueNDRangeKernel):
> _If local_work_size is specified, the values specified in global_work_size[0], …​, global_work_size[work_dim - 1] must be evenly divisible by the corresponding values specified in local_work_size[0], …​, local_work_size[work_dim - 1]._

The above implies that `global_work_size[n]` must be _at least_ divisible by `local_work_size[n]`.

- This PR fixes the corner case in Im3d2Col when `global_work_size[0]` is so small that it becomes smaller than `local_work_size[0]`.
  - Other cases  (like potentially uneven divisibility) are outside the scope of this fix as they are expected to be handled by the existing code.
- The PR also proactively fixes the same potential issue for the rest of col2im/im2col and layout transposing kernels.
- Regression test added.

### Tested on Navi21 and Radeon VII, with FP32, FP16, BF16 and INT8.

----
[Attribution] @junliume @johnny-keker https://github.com/ROCmSoftwarePlatform/MIOpen/labels/workaround https://github.com/ROCmSoftwarePlatform/MIOpen/labels/urgency_high
- Proposed reviewers: @JehandadKhan @junliume @zjing14 
- Milestone: ROCm 5.6

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -650,41 +650,49 @@ if(${CODECOV_TEST})
...
         #   https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1935
-        COMMAND MIOPEN_DRIVER_USE_GPU_REFERENCE=1 $<TARGET_FILE:MIOpenDriver> convfp16 -n 2 -c 64 --in_d 128 -H 128 -W 128 -k 32 --fil_d 3 -y 3 -x 3 --pad_d 1 -p 1 -q 1 --conv_stride_d 1 -u 1 -v 1 --dilation_d 1 -l 1 -j 1 --spatial_dim 3 -m conv -g 1 -F 1 -t 1
+        COMMAND MIOPEN_DRIVER_USE_GPU_REFERENCE=1 $<TARGET_FILE:MIOpenDriver> ${MIOPENDRIVER_MODE_CONV} -n 2 -c 64 --in_d 128 -H 128 -W 128 -k 32 --fil_d 3 -y 3 -x 3 --pad_d 1 -p 1 -q 1 --conv_stride_d 1 -u 1 -v 1 --dilation_d 1 -l 1 -j 1 --spatial_dim 3 -m conv -g 1 -F 1 -t 1
     )
 
-    set_tests_properties(test_miopendriver_big_tensor
-        PROPERTIES RUN_SERIAL On)
-    set_tests_properties(test_miopendriver_big_tensor
-        PROPERTIES TIMEOUT 600)
+    add_custom_test(test_miopendriver_regression GFX103X_ENABLED GFX110X_ENABLED HALF_ENABLED BF16_ENABLED INT8_ENABLED
```
Comment (junliume) [inline test/CMakeLists.txt]: @atamazov this test failed in CI, hence `-DBUILD_DEV=On` is included in cmake command for this stage: Fp32 OpenCL Debug + Codec…

[fail.log](https://github.com/ROCmSoftwarePlatform/MIOpen/files/11094265/fail.log)

Comment (atamazov) [reply test/CMakeLists.txt]: Thanks. Resolved in [e45aef8](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2052/commits/e45aef8ad07da157213dcd5b723e4433c8e4b3bd)

**Comment by atamazov** [comment]
@junliume @johnny-keker Any CI issues?

**Comment by junliume** [comment]
Hi @atamazov  we have updated CI docker, could you please merge the latest develop branch into your branch? It should resolve these issues related to compiler update.

**Comment by atamazov** [comment]
Hi @junliume done!

**Review by JehandadKhan — APPROVED** [review]

**Review by junliume — APPROVED** [review]

**Comment by junliume** [comment]
Ping @zjing14 and @carlushuang  for last check

**Comment by junliume** [comment]
@atamazov verification of this fix is working on not on the latest mainline is blocked by https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2062

**Comment by junliume** [comment]
verified after https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2063

**Pull request merged**